File: cwsfroster/MembersTableViewController.m
//
//  MembersTableViewController.m
//  cwsfroster
//
//  Created by Bobby Ren on 5/28/14.
//  Copyright (c) 2014 Bobby Ren. All rights reserved.
//

#import "MembersTableViewController.h"

@interface MembersTableViewController ()

@end

@implementation MembersTableViewController

- (id)initWithStyle:(UITableViewStyle)style
{
    self = [super initWithStyle:style];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    [button setImage:[UIImage imageNamed:@"hamburger4-square"] forState:UIControlStateNormal];
    [button setFrame:CGRectMake(0, 0, 30, 30)];
    [button addTarget:self action:@selector(goToSettings:) forControlEvents:UIControlEventTouchUpInside];
    UIBarButtonItem *left = [[UIBarButtonItem alloc] initWithCustomView:button];
    self.navigationItem.leftBarButtonItem = left;

    UIButton *rightbutton = [UIButton buttonWithType:UIButtonTypeCustom];
    [rightbutton setImage:[UIImage imageNamed:@"plus"] forState:UIControlStateNormal];
    [rightbutton setFrame:CGRectMake(0, 0, 30, 30)];
    [rightbutton addTarget:self action:@selector(goToAddMember:) forControlEvents:UIControlEventTouchUpInside];
    UIBarButtonItem *right = [[UIBarButtonItem alloc] initWithCustomView:rightbutton];
    self.navigationItem.rightBarButtonItem = right;

    [self listenFor:@"payment:updated" action:@selector(reloadMembers)];
    [self reloadMembers];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

-(void)reloadMembers {
    [self.tableView reloadData];
}

-(void)goToSettings:(id)sender {
    [self notify:@"goToSettings"];
}

-(void)goToAddMember:(id)sender {
    [self performSegueWithIdentifier:@"toAddMember" sender:nil];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return [[[Organization current] members] count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    MemberCell *cell = [tableView dequeueReusableCellWithIdentifier:@"MemberCell" forIndexPath:indexPath];
    
    // Configure the cell...

    Member *member = [[Organization current] members][indexPath.row];
    [cell configureWithMember: member row:indexPath.row];
    
    return cell;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return YES if you want the specified item to be editable.
    return YES;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        //add code here for when you hit delete
        [self deleteMemberAtIndexPath:indexPath];
    }
}

#pragma mark - Navigation
// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    UINavigationController *nav = segue.destinationViewController;
    MemberInfoViewController *controller = (MemberInfoViewController *)(nav.topViewController);
    if ([segue.identifier isEqualToString:@"toEditMember"]) {
        Member *member = [[Organization current] members][[self.tableView indexPathForSelectedRow].row];
        [controller setDelegate:self];
        [controller setMember:member];
    }
    else if ([segue.identifier isEqualToString:@"toAddMember"]) {
        [controller setDelegate:self];
        [controller setMember: nil];
    }
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}

#pragma mark Delegate
-(void)didUpdateMember:(Member *)member {
    [self reloadMembers];
    [self notify:@"member:updated"];
    [ParseLog logWithTypeString:@"MemberUpdated" title:[member objectId] message:nil params:nil error:nil];
}

-(void)close {
    [self.navigationController dismissViewControllerAnimated:YES completion:nil];
}

-(void)deleteMemberAtIndexPath:(NSIndexPath *)indexPath {
    Member *member = [[Organization current] members][indexPath.row];
    [member deleteInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
        if (succeeded) {
            [Organization queryForMembersWithCompletion:^(NSArray<Member *> * _Nullable members, NSError * _Nullable error) {
                [self.tableView reloadData];
                [self notify:@"member:deleted"];
                [ParseLog logWithTypeString:@"MemberDeleted" title:[member objectId] message:nil params:nil error:nil];
            }];
        }
    }];
}
@end

File: cwsfroster/MembersTableViewController.swift
//
//  MembersTableViewController.swift
//  rollcall
//
//  Created by Bobby Ren on 6/10/18.
//  Copyright © 2018 Bobby Ren. All rights reserved.
//

import UIKit

protocol MemberDelegate: class {
    func didCreateMember(_ member: FirebaseMember)
    func didUpdateMember(_ member: FirebaseMember)
}

class MembersTableViewController: UITableViewController {
    fileprivate var _members: [FirebaseMember] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()

        listenFor("payment:updated", action: #selector(reloadMembers), object: nil)
        listenFor("member:created", action: #selector(reloadMembers), object: nil)

        setupSettingsNavButton()
        setupPlusNavButton()
        reloadMembers()
    }
    
    fileprivate func setupSettingsNavButton() {
        let button = UIButton(frame: CGRect(x: 0, y: 0, width: 30, height: 30))
        button.setImage(UIImage(named: "hamburger4-square"), for: .normal)
        button.addTarget(self, action: #selector(goToSettings), for: .touchUpInside)
        navigationItem.leftBarButtonItem = UIBarButtonItem(customView: button)
    }
    
    fileprivate func setupPlusNavButton() {
        let button = UIButton(frame: CGRect(x: 0, y: 0, width: 30, height: 30))
        button.setImage(UIImage(named: "plus"), for: .normal)
        button.addTarget(self, action: #selector(goToAddMembers), for: .touchUpInside)
        navigationItem.rightBarButtonItem = UIBarButtonItem(customView: button)
    }
    
    @objc func reloadMembers() {
        OrganizationService.shared.members { [weak self] (members, error) in
            self?._members = members.sorted{
                guard let n1 = $0.name?.uppercased() else { return false }
                guard let n2 = $1.name?.uppercased() else { return true }
                return n1 < n2
            }
            self?.tableView.reloadData()
        }
    }
    
    @objc func goToSettings() {
        notify("goToSettings", object: nil, userInfo: nil)
    }
    
    @objc func goToAddMembers() {
        performSegue(withIdentifier: "toAddMembers", sender: nil)
    }
    
    fileprivate func deleteMember(at row: Int) {
        guard row < _members.count else { return }
        let member = _members[row]
        OrganizationService.shared.deleteMember(member) { [weak self] (success, error) in
            self?.reloadMembers()
            if !success {
                self?.simpleAlert("Could not delete member", message: "There was an issue deleting this member. No changes have been made.")
            }
        }
    }
    
    func close() {
        navigationController?.dismiss(animated: true, completion: nil)
    }
}

extension MembersTableViewController {
    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return _members.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell: MemberCell = tableView.dequeueReusableCell(withIdentifier: "MemberCell", for: indexPath) as? MemberCell else { return UITableViewCell() }

        let row = indexPath.row
        guard row < _members.count else { return cell }
        cell.configure(member: _members[row], row: row)

        return cell
    }

    // Override to support conditional editing of the table view.
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        // Return false if you do not want the specified item to be editable.
        return true
    }

    // Override to support editing the table view.
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            // Delete the row from the data source
            deleteMember(at: indexPath.row)
            reloadMembers()
        }
    }
}

extension MembersTableViewController {
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "toEditMember", let row = tableView.indexPathForSelectedRow?.row {
            guard let nav = segue.destination as? UINavigationController, let controller = nav.viewControllers.first as? MemberInfoViewController else { return }
            controller.delegate = self
            let member = _members[row]
            controller.member = member
        } else if segue.identifier == "toAddMembers" {
            guard let nav = segue.destination as? UINavigationController, let _ = nav.viewControllers.first as? AddMembersViewController else { return }
//            controller.delegate = self
        }
    }
}

extension MembersTableViewController: MemberDelegate {
    func didCreateMember(_ member: FirebaseMember) {
        _members.append(member)
        _members = _members.sorted {
            guard let n1 = $0.name?.uppercased() else { return false }
            guard let n2 = $1.name?.uppercased() else { return true }
            return n1 < n2
        }
        tableView.reloadData()
        notify("member:updated", object: nil, userInfo: nil)
    }
    
    func didUpdateMember(_ member: FirebaseMember) {
        tableView.reloadData()
        notify("member:updated", object: nil, userInfo: nil)
    }
}
